### PR TITLE
Fix module resolution for non standard filenames

### LINKED ISSUE #668: Module resolution fails when using non-standard filenames.
When using a combination of directory modules and a name other than lib.rs/main.rs module resolution fails.

Cargo allows you to have a name other than e.g. lib.rs for libraries like this.

```toml
[package]
...

[lib]
path = "src/my_library.rs"
``` 

Relevant code:
https://github.com/rust-analyzer/rust-analyzer/blob/2acaa92c93bcd54d97221d0cede8780ff817476e/crates/ra_hir/src/module_tree.rs#L291-L306

### PR DESCRIPTION
**Pull request by regiontog**
fixes #668

### CODE REVIEW

File: crates/ra_hir/src/nameres/tests.rs [hunk 1]
```diff
@@ -19,6 +19,19 @@ fn item_map(fixture: &str) -> (Arc<ItemMap>, ModuleId) {
     (db.item_map(krate.crate_id), module_id)
 }
 
+fn item_map_custom_crate_root(fixture: &str, root: &str) -> (Arc<ItemMap>, ModuleId) {
+    let (mut db, pos) = MockDatabase::with_position(fixture);
+
+    let mut crate_graph = CrateGraph::default();
+    crate_graph.add_crate_root(db.file_id(root));
```
Comment (matklad) [inline crates/ra_hir/src/nameres/tests.rs]: I think this is the same as `pos.file_id`? That is, you don't need to add `fn file_id` to MockDB. In fact, `FileMap` is a dead code, I should remove it :)

Comment (regiontog) [reply crates/ra_hir/src/nameres/tests.rs]: Ah, that makes it easier, the FileMap was added by me. MockDatabase just used a incremental counter.

Comment (regiontog) [reply crates/ra_hir/src/nameres/tests.rs]: On second thought that would require you to leave the position marker on the crate root I think.

Comment (matklad) [reply crates/ra_hir/src/nameres/tests.rs]: but that what happens anyway?

https://github.com/rust-analyzer/rust-analyzer/pull/690/files#diff-2e7269539c8928d62c4fb71200a3a2a9R158

Comment (regiontog) [reply crates/ra_hir/src/nameres/tests.rs]: That is, true. I suppose I might be a conflict if someone wants to use `item_map_custom_crate_root` later. If you prefer to use the marker position I can document that in the function though.

Comment (matklad) [reply crates/ra_hir/src/nameres/tests.rs]: Yeah, let's just use a marker position.

In my experience, such "test setup functions" always feel a little bit ad-hock and non-orthogonal, but it do not cause much problems in practice. 

If we'd like, we can get the current behavior of mathcing by file name by iterating through source roots, though I'd reuse existing cursor marker infra

**Comment by matklad** [comment]
bors r+

Thanks!

**Pull request merged**